src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
 % The Stack and the Heap
 
-Coming Soon
+As a systems language, Rust operates at a low level. If you’re coming from a
+high-level language, there’s some aspects of systems programming that you may
+not be familiar with. The most important one is how memory works, with a stack
+and a heap. If you’re familiar with the terms ‘stack’ and ‘heap’, this chapter
+will be a refresher. If you’re not, you’ll learn about this more general
+concept, but with a Rusty focus.
+
+# Memory management
+
+These two terms are about memory management. The stack and the heap are
+abstractions that help you determine when to allocate and deallocate memory.
+
+Here’s a high-level comparison:
+
+The stack is very fast, and is where memory is allocated in Rust by default.
+But the allocation is local to a function call, and is limited in size. The
+heap, on the other hand, is slower, and is explicitly allocated by your
+program. But it’s effectively unlimited in size, and is globally accessible.
+
+# The Stack
+
+Let’s talk about this Rust program:
+
+```rust
+fn main() {
+    let x = 42;
+}
+```
+
+This program has one variable binding, `x`. This memory needs to be allocated
+from somewhere. Rust ‘stack allocates’ by default, which means that basic
+values ‘go on the stack’. What does that mean?
+
+Well, when a function gets called, some memory gets allocated for all of its
+local variables and some other information. This is called a ‘stack frame’, and
+for the purpose of this tutorial, we’re going to ignore the extra information
+and just consider the local variables we’re allocating. So in this case, when
+`main()` is run, we’ll allocate a single 32-bit integer for our stack frame.
+This is automatically handled for you, as you can see, we didn’t have to write
+any special Rust code or anything.
+
+When the function is over, its stack frame gets deallocated. This happens
+automatically, we didn’t have to do anything special here.
+
+That’s all there is for this simple program. The key thing to understand here
```

Comment:
In all major OSes we support stack is fast mostly not because we ”grab the memory all at once” for a function, but because each thread already starts with a big chunk of static memory allocated, which is what we call a stack, therefore we don’t even need to ask OS to do an actual allocation or deallocation when a stack frame is pushed or popped.